src/main/java/bio/terra/service/snapshotbuilder/SnapshotBuilderService.java

```diff
@@ -206,51 +208,75 @@ private EnumerateSnapshotAccessRequest convertToEnumerateModel(
-    return concepts.stream().filter(c -> c.getId().equals(id)).findFirst().orElseThrow();
+    ParentQueryResult(FieldValueList row) {
+      this(
+          (int) row.get(HierarchyQueryBuilder.PARENT_ID).getLongValue(),
+          (int) row.get(HierarchyQueryBuilder.CONCEPT_ID).getLongValue(),
+          row.get(HierarchyQueryBuilder.CONCEPT_NAME).getStringValue());
+    }
   }
 
-  // Map of concept id to parent concept id.
-  private static final Map<Integer, Integer> HIERARCHY =
-      Map.of(406, 404, 405, 404, 404, 401, 403, 401, 407, 402, 401, 400, 400, 100, 408, 100);
-
   public SnapshotBuilderGetConceptHierarchyResponse getConceptHierarchy(
-      UUID id, int conceptId, AuthenticatedUserRequest userRequest) {
-    var concepts = loadConcepts();
-    SnapshotBuilderConcept concept = getConcept(concepts, conceptId);
-    while (true) {
-      // For a child concept, generate its parent concept and siblings until we reach the root.
-      var parentId = HIERARCHY.get(concept.getId());
-      if (parentId == null) {
-        break;
-      }
-      var parent = getConcept(concepts, parentId);
-      concept =
-          parent.children(
-              HIERARCHY.entrySet().stream()
-                  .filter(e -> e.getValue().equals(parentId))
-                  .map(entry -> getConcept(concepts, entry.getKey()))
-                  .toList());
+      UUID datasetId, int conceptId, AuthenticatedUserRequest userRequest) {
+    Dataset dataset = datasetService.retrieve(datasetId);
+    var query =
+        queryBuilderFactory
+            .hierarchyQueryBuilder(getTableNameGenerator(dataset, userRequest))
+            .generateQuery(conceptId);
+    var sql = query.renderSQL(CloudPlatformWrapper.of(dataset.getCloudPlatform()));
+
+    Map<Integer, SnapshotBuilderParentConcept> parents = new HashMap<>();
+    runSnapshotBuilderQuery(sql, dataset, ParentQueryResult::new, ParentQueryResult::new)
+        .forEach(
+            row -> {
+              SnapshotBuilderParentConcept parent =
+                  parents.computeIfAbsent(
+                      row.parentId,
+                      k ->
+                          new SnapshotBuilderParentConcept()
+                              .parentId(k)
+                              .children(new ArrayList<>()));
+              parent.addChildrenItem(
+                  new SnapshotBuilderConcept()
+                      .id(row.childId)
+                      .name(row.childName)
+                      .hasChildren(true)
+                      .count(1));
+            });
+    if (parents.isEmpty()) {
+      throw new BadRequestException("No parents found for concept %s".formatted(conceptId));
     }
-    return new SnapshotBuilderGetConceptHierarchyResponse().result(concept);
+
+    return new SnapshotBuilderGetConceptHierarchyResponse().result(moveRootToFirst(parents));
+  }
+
+  /** Given a list of hierarchy parents, find the root and move it to the first entry. */
```

Comment:
Thanks for the comments! I understand what this is doing, but I'm wondering why this is needed? Also can there be multiple root parents?

Reply:
This behavior is part of the API contract; the first parent returned is the root. I could change it so the root is re-derived by the UI code though.

Reply:
I've changed the code so the UI is now deriving the root. If we limit the response to standard concepts only, there should only be one unique root. But we may need to revisit this later on depending the data we end up seeing.